bindings/python/convert.py

```diff
@@ -1,72 +1,197 @@
-        with open(filename, "r") as f:
-            data = json.load(f)
-
-        filenames = set(data["weight_map"].values())
-        for filename in filenames:
-            cached_filename = hf_hub_download(repo_id=model_id, filename=filename)
-            loaded = torch.load(cached_filename)
-            local = rename(filename)
-            save_file(loaded, local, metadata={"format": "pt"})
-            local_filenames.append(local)
-
-        index = "model.safetensors.index.json"
-        with open(index, "w") as f:
-            newdata = {k: v for k, v in data.items()}
-            newmap = {k: rename(v) for k, v in data["weight_map"].items()}
-            newdata["weight_map"] = newmap
-            json.dump(newdata, f)
-        local_filenames.append(index)
-
-        api = HfApi()
-        operations = [CommitOperationAdd(path_in_repo=local, path_or_fileobj=local) for local in local_filenames]
-        api.create_commit(
-            repo_id=model_id,
-            operations=operations,
-            commit_message="Adding `safetensors` variant of this model",
-            create_pr=True,
-        )
-    finally:
-        for local in local_filenames:
-            os.remove(local)
-
-
-def convert_single(model_id):
-    local = "model.safetensors"
-    try:
-        filename = hf_hub_download(repo_id=model_id, filename="pytorch_model.bin")
-        loaded = torch.load(filename)
+def convert_multi(model_id: str) -> List["CommitOperationAdd"]:
+    filename = hf_hub_download(repo_id=model_id, filename="pytorch_model.bin.index.json")
+    with open(filename, "r") as f:
+        data = json.load(f)
+
+    filenames = set(data["weight_map"].values())
+    for filename in filenames:
+        cached_filename = hf_hub_download(repo_id=model_id, filename=filename)
+        loaded = torch.load(cached_filename)
+        sf_filename = rename(filename)
+
+        local = os.path.join(folder, sf_filename)
         save_file(loaded, local, metadata={"format": "pt"})
+        check_file_size(local, cached_filename)
+        local_filenames.append(local)
+
+    index = os.path.join(folder, "model.safetensors.index.json")
+    with open(index, "w") as f:
+        newdata = {k: v for k, v in data.items()}
+        newmap = {k: rename(v) for k, v in data["weight_map"].items()}
+        newdata["weight_map"] = newmap
+        json.dump(newdata, f)
+    local_filenames.append(index)
+
+    operations = [CommitOperationAdd(path_in_repo=local.split("/")[-1], path_or_fileobj=local) for local in local_filenames]
+
+    return operations
+
+
+def convert_single(model_id: str, folder: str) -> List["CommitOperationAdd"]:
+    sf_filename = "model.safetensors"
+    filename = hf_hub_download(repo_id=model_id, filename="pytorch_model.bin")
+    loaded = torch.load(filename)
+
+    local = os.path.join(folder, sf_filename)
+    shared = shared_pointers(loaded)
+    for shared_weights in shared:
+        for name in shared_weights[1:]:
+            loaded.pop(name)
```

Comment:
I don't think it's a good idea to delete those from the state dict. It will definitely surprise users using `safetensors` independently from Transformers, and even in Transformers this will lead to new warnings for users and possible failures as I think a good half of the models don't have tied weights properly implemented.

Reply:
This is checked against during the reload, so the PR won't be opened if that's the case.

I found FSMT in the top 100 that's impacted.

Duplicating the weights in disk WILL crash fine-tuning anyway no ?
So it would be better to implement tying weights, no ? (And crash early when it isn't .)

Reply:
I don't think you are checking against new warnings.

In both cases I don't see where there is going to be crash. In one instance (weights duplicated) inference will work and fine-tuning will have disconnected weights, in the other (weights not duplicated) inference will not work (since the weights won't be found) and fine-tuning will have disconnected weights.

In any case I would trust the PyTorch state dict as the absolute source of trust.